test/remediation.test.ts

```diff
@@ -0,0 +1,68 @@
+import { test } from 'tap';
+import fs = require('fs');
+
+import plugin = require('../lib');
+import { chdirWorkspaces } from './test-utils';
+const testUtils = require('./test-utils') as any;
+
+test('remediation with constraints.txt', async (t) => {
+  chdirWorkspaces('pip-app');
+  t.teardown(testUtils.activateVirtualenv('pip-app'));
+  const upgrades = {
+    'Django@1.6.1': { upgradeTo: 'Django@2.0.1' },
+    'transitive@1.0.0': { upgradeTo: 'transitive@1.1.1' },
+  };
+  const manifests = {
+    'requirements.txt': fs.readFileSync('requirements.txt', 'utf8'),
+    'constraints.txt': '',
+  };
+
+  const expectedUpdatedManifests = {
+    'requirements.txt':
+      'Jinja2==2.7.2\ndjango==2.0.1\npython-etcd==0.4.5\nDjango-Select2==6.0.1 # this version installs with lowercase so it catches a previous bug in pip_resolve.py\nirc==16.2 # this has a cyclic dependecy (interanl jaraco.text <==> jaraco.collections)\ntesttools==\\\n    2.3.0 # this has a cycle (fixtures ==> testtols);\n./packages/prometheus_client-0.6.0',
+    'constraints.txt':
+      '\ntransitive>=1.1.1 # pinned by Snyk to avoid a vulnerability\n',
+  };
+
+  const result = await plugin.applyRemediationToManifests(
+    '.',
+    manifests,
+    upgrades,
+    {}
+  );
+
+  t.same(result, expectedUpdatedManifests, 'remediation as expected');
+});
+
+test('remediation without constraints.txt', async (t) => {
```

Comment:
How about tests that can confirm/assert expected behaviour when unexpected things happen:
- empty upgrades object is passed?
- a computed upgrade is actually a `downgrade`
- do we know to remediate system specific packages? Like windows or mac specific ones?
- how does Python 2 / 3 play a role if at all in this remediation?

Reply:
added a code to retain known "extra markers" during application of remediation and a test for one of the cases.